pingcap/tidb
# txn: fix the inconsistency index keys caused by transaction amend for add index

### What problem does this PR solve?

Issue Number: 
Solve #20899 in release-4.0 branch.
Solve #20910 in release-4.0 branch

Problem Summary:

For `20899`
The inconsistency caused by transaction amend for add index ddl change. For such scenerios
```
init   create table t1(c1 int key, c2 int, c3 int, c4 int);
txn1 begin
txn2 insert into t1 values(1, 10, 100, 111);
txn2 alter table t1 add unique key uk(c2);
txn2 alter table t1 add index k1(c3);
txn1 update t1 set c4 = c4 + 1; 
txn1 commit; 
```
The old implementations will generate these keys for amend prewrite 
```
Op_Del   index_key_for_uk -> nil 
Op_Put   index_key_for_uk -> 1
Op_Del   index_key_for_k1_1 -> v
Op_Put   index_key_for_k1_1 -> v
```
If the prewrite batch limit is hit, the delete and put mutations for the same key are split into different prewrte requests and the execution sequence is not kept, the old index key may be deleted. Actually for the untouched index keys it's not needed.


For `20910`
We should not rely on write conflict error to report conflict unique index value, but force to check existency.

### What is changed and how it works?

What's Changed:

For `20899`
Merge the `delete` and `add` operations into single `amendOperationAddIndex`. Check if the genereted index keys from old row value and new row value are the same if "delete" and "insert" are both needed for processing a single row. If the generated index keys are the same, skip them and do not add them into result mutations, thus there will not be duplicate prewrite requests having
both `Op_Del` and `Op_put` for a single index key.

For `20910`
Add the forced existency assertions in `handlePessimisticAction` for processing amend prewrites.

Besides, there should be **no** duplicate key entries in generated amend mutations, so add this check at the end of the amend process.

How it Works:

### Related changes


### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)
  Use the case generated by @you06
   Run update statements which may update row value, check if the generated amend prewrite requests, they should not contain mutations related to untouched index keys.

Side effects



### Release note 

- Fix the duplicate amend prewrite requests which may cause inconsistency in index keys.
